docs/BinaryProtocol.md

```diff
@@ -0,0 +1,565 @@
+
+# Pulsar binary protocol specification
+
+<!-- TOC depthFrom:2 depthTo:3 withLinks:1 updateOnSave:1 orderedList:0 -->
+
+- [Framing](#framing)
+- [Message metadata](#message-metadata)
+	- [Batch messages](#batch-messages)
+- [Interactions](#interactions)
+	- [Connection establishment](#connection-establishment)
+	- [Keep Alive](#keep-alive)
+	- [Producer](#producer)
+	- [Consumer](#consumer)
+- [Service discovery](#service-discovery)
+	- [Topic lookup](#topic-lookup)
+	- [Partitioned topics discovery](#partitioned-topics-discovery)
+
+<!-- /TOC -->
+
+Pulsar uses a custom binary protocol for communications between
+producers/consumers and brokers. The protocol was design to support all the
+required features (eg: acknowledgments, flow control) while ensuring maximum
+transport and implementation efficiency.
+
+Clients and brokers exchange *commands* between each other. Commands are
+formatted as binary [Protocol Buffers](https://developers.google.com/protocol-buffers/)
+messages.
+
+The format of the protobuf commands is specified in the `PulsarApi.proto`
+file
+[pulsar-common/src/main/proto/PulsarApi.proto](https://github.com/yahoo/pulsar/blob/master/pulsar-common/src/main/proto/PulsarApi.proto)
+
+Commands for different producers and consumers can be sent through the same
+connection, without any restriction to the interleaving of such commands.
+
+
+
+All the commands are embedded in a [`BaseCommand`](https://github.com/yahoo/pulsar/blob/master/pulsar-common/src/main/proto/PulsarApi.proto#L283) protobuf object that includes a type enum and all the possible commands as
+optional fields. At any time, one `BaseCommand` can only have set a single
```

Comment:
Should we call it BaseCommand and SubCommand? Then it would be:
All the commands are embedded in a BaseCommand protobuf object that includes a type enum and all possible SubCommands as optional field. At any given time, one BaseCommand can only have one SubCommand.